torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -240,17 +241,21 @@ void FusionExecutor::compileFusion(
 
   // TODO: pass block_size here;
   c10::optional<int> block_size = c10::nullopt;
-  if (!inputs.empty()) {
-    auto expr_eval = executor_utils::bindKernelInputs(inputs, kernel);
+  if (!args.empty()) {
+    auto expr_eval = executor_utils::bindKernelInputs(args, kernel);
     auto launch_params =
         computeLaunchParams(launch_constraints, expr_eval, warp_size_);
     block_size = launch_params.nThreads();
     TORCH_INTERNAL_ASSERT(
         block_size > 0, "launch param inferred block size < 0");
   }
 
-  block_size_high_water_mark =
-      block_size.has_value() ? block_size.value() : block_size_high_water_mark;
+  // Basically setting high water martk as 1 when we don't provide args for
+  // compilation, it will just generate a kernel that gets ditched at the first
+  // run - not great. We should have better heuristics.
```

Comment:
https://developer.nvidia.com/blog/cuda-pro-tip-occupancy-api-simplifies-launch-configuration/

^^^ I'll put a todo there so we can follow up and refactor how we handle high water mark. We also need to plumb this to runtime heuristics instead of silently trigger re-compilation inside kernel execution. I suspect that's already done but don't know for sure yet. cc'ing @shmsong